Client/Frontend/Browser/Authenticator.swift

```diff
@@ -48,6 +48,48 @@ class Authenticator {
         return self.promptForUsernamePassword(viewController, credentials: nil, protectionSpace: challenge.protectionSpace, loginsHelper: nil)
     }
 
+    static func findMatchingCredentialsForChallenge(challenge: NSURLAuthenticationChallenge, fromLoginsProvider loginsProvider: BrowserLogins) -> Deferred<NSURLCredential?> {
+        let deferred = Deferred<NSURLCredential?>()
```

Comment:
If you make this method return `Deferred<Maybe<NSURLCredential?>>` then you have two possible ways of expressing failure: a failed `Maybe` for an error, and a `nil` credential for no match in the DB. And you can do this without manually futzing with `Deferred` :)